Samples/Raytracing/PathTracer/Data/GGXGICommon.slang

```diff
@@ -0,0 +1,224 @@
+    int lightIndex = min(int(rand_next(rndSeed) * gLightsCount), gLightsCount - 1);
+
+    LightData ld = gLights[lightIndex];
+    ShadingResult sr = evalMaterial(sd, gLights[lightIndex], 1.0f);
+
+    // Shoot our shadow ray to our randomly selected light
+    float distToLight = length(ld.posW - sd.posW);
+    float3 L = -normalize(ld.dirW);
+    float shadowMult = float(gLightsCount) * shootShadowRay(sd.posW, L, gMinT, distToLight);
+
+    return shadowMult * sr.color.rgb;
+}
+
+/** Helper for shooting a ray for indirect lighting. Assumes hit group #1 and miss shader #1
+
+    \return Color sampled by the ray.
+*/
+float3 shootIndirectRay(float3 rayOrigin, float3 rayDir, float minT, uint curPathLen, uint seed, uint curDepth)
+{
+    // Setup our indirect ray
+    RayDesc rayColor;
+    rayColor.Origin = rayOrigin;
+    rayColor.Direction = rayDir;
+    rayColor.TMin = minT;
+    rayColor.TMax = 1.0e38f;
+
+    // Initialize the ray's payload data with black return color and the current rng seed
+    IndirectRayPayload payload;
+    payload.color = float3(0, 0, 0);
+    payload.rndSeed = seed;
+    payload.rayDepth = curDepth + 1;
+
+    // Trace our ray to get a color in the indirect direction.  Use hit group #1 and miss shader #1
+    TraceRay(gRtScene, 0, 0xFF, 1, hitProgramCount, 1, rayColor, payload);
+
+    // Return the color we got from our ray
+    return payload.color;
+}
+
+/** Our material has have both a diffuse and a specular lobe.
+    Calculates what probability the diffuse one should be sampled.
+
+    \return Probability between 0 and 1
+*/
+float probabilityToSampleDiffuse(float3 difColor, float3 specColor)
+{
+    float lumDiffuse = max(0.01f, luminance(difColor.rgb));
+    float lumSpecular = max(0.01f, luminance(specColor.rgb));
+    return lumDiffuse / (lumDiffuse + lumSpecular);
+}
+
+/** Evaluates indirect lighting up to a bounce count specified by gMaxDepth. Initial call should leave rayDepth = 0.
+*/
+float3 ggxIndirect(inout uint rndSeed, ShadingData sd, uint rayDepth = 0)
+{
+    // We have to decide whether we sample our diffuse or specular/ggx lobe.
+    float probDiffuse = probabilityToSampleDiffuse(sd.diffuse, sd.specular);
+    float chooseDiffuse = (rand_next(rndSeed) < probDiffuse);
+    float2 randVal = float2(rand_next(rndSeed), rand_next(rndSeed));
+
+    // We'll need NdotV for both diffuse and specular...
+    float NdotV = saturate(dot(sd.N, sd.V));
+
+    // If we randomly selected to sample our diffuse lobe...
+    if (chooseDiffuse)
+    {
+        // Shoot a randomly selected cosine-sampled diffuse ray.
+        float3 L = getCosHemisphereSample(randVal, sd.N, getPerpendicularStark(sd.N));
+        float3 bounceColor = shootIndirectRay(sd.posW, L, gMinT, 0, rndSeed, rayDepth);
+
+        // Check to make sure our randomly selected, normal mapped diffuse ray didn't go below the surface.
+        if (dot(sd.N, L) <= 0.0f) bounceColor = float3(0, 0, 0);
```

Comment:
I don't think this does what you think it does.  sd.N stores the shading normal.  `getCosHemisphereSample` and similar functions should *never* return a value with N dot L < 0 (at least if properly written).  However, if you have the geometric normal, N_g dot L can be < 0, due to normal/bump mapping.  Basically, if your normal map perturbs your normal by 90 degrees in one direction, half of your random vectors will be below the surface.  This is what currently causes the blue light leaking on the wainscoting in the Arcade room.   (Basically, as written, this line of code does nothing unless you use your pre-normal mapped normal and you had passed a different normal to `getCosHemisphereSample`)